File: MaxMulti.java
// Give a O(n) algorithm to compute the max product of a successive sub array


public class MaxMulti {
	public static double Max_Num(double[] a) {
		double max_positive = 1;
		double min_negative = 1;
		double max = 1;
		
		for (int i = 0; i < a.length; i++) {
			if (a[i] > 0) {
				max_positive = max_positive *a[i];
				min_negative = Math.min(min_negative*a[i], 1);
			}else if(a[i] == 0) {
				max_positive = 1; 
				min_negative = 1;
			}else {
				double temp = max_positive;
				max_positive = Math.max(min_negative*a[i], 1);
				min_negative = temp*a[i];
			}
			max = Math.max(max, max_positive);
		}
		return max;
	}
	
	public static void main(String[] args) {
		double[] a = {1.5, 3, -2, 2.5, 10.2, 0, 12, -3};
		System.out.println(Max_Num(a));
	}
}

File: algorithm/max_multi.cc
/* 给一个长度为N的整数数组，
 * 只允许用乘法，不能除法，
 * 计算任意N-1个数的乘积最大值
 */

/* 用两个额外的数组
 * f[i] 表示前i个元素的成绩
 * b[i] 表述后i个元素的乘积
 *
 * 则e[i] 表示除去i 之外其他n-1
 * 个元素的乘积
 *
 * 则 e[i] = f[i-1] * b[i+1];
 */
#include <climits>
#include <cstdio>
int maxmulti(int *a, int n) {
    int *f = new int[n+1];
    int *b = new int[n+2];
    f[0] = 1;
    b[n+1] = 1;

    for (int i = 1; i <= n; i++) {
        f[i] = f[i-1] * a[i-1];
        b[n-i+1] = b[n-i+2] * a[n-i];
    }

    int max =  INT_MIN;
    for (int i = 1; i <= n; i++) {
        int temp = f[i-1] * b[i+1];
        if ( max < temp)
            max = temp;
    }

    delete[] f;
    delete[] b;
    return max;
}

int main() {
    int a[] = {1,5,4,2,-7};
    int max = maxmulti(a, 5);
    printf("%d\n", max);
    return 0;
}

